ldm/simplet2i.py

```diff
@@ -311,7 +315,10 @@ def process_image(image,seed):
                     callback=step_callback,
                 )
 
-            with scope(self.device.type), self.model.ema_scope():
+            device_type = self.device.type # this returns 'mps' on M1
+            if device_type != 'cuda' or device_type != 'cpu':
+                device_type = 'cpu'
+            with scope(device_type), self.model.ema_scope():
```

Comment:
autocast seems to only support `cuda` and `cpu` as options here, otherwise it throws an exception

Reply:
Take a look at [this](https://github.com/lstein/stable-diffusion/blob/main/ldm/dream/devices.py)

Reply:
yes, I'm using `choose_torch_device()` in `__init__` and using the string it returns (cuda, mps, cpu) to pass into `torch.device()` to get an object representing the device. What is going on here in this piece of code is that `device.type` returns a string of (cuda, mps, cpu) which was being passed into autocast(), however autocast does not recognize mps as valid, so for mps I need to pass in cpu as the device type.

Reply:
Duh, sorry, I thought `__init_` was a file but now I realize it's a subroutine.

Reply:
Ok, I'm catching up with you. In txt2img.py we added this code for the same effect:
```
    if device.type in ['mps', 'cpu']:
        precision_scope = nullcontext # have to use f32 on mps
```
I'm just wondering if there is a better way to do this. Maybe this logic should be in ldm/dream/device.py also?

Reply:
I don't understand enough about how this all works to make a call on what the best way to handle this is. I just tried passing `cpu` and it worked so...

Reply:
I'm not really up on the torch devices, but in general all the code that detects the environment and selects the appropriate device should go into one place. I created ldm.dream.devices.choose_torch_device() for this, but it's just a placeholder and I'm happy to see it being turned into a class that we can inherit from.

Reply:
I'll make a special function to choose an autocast device. Just so we're all clear, this is specific to autocast because for some reason beyond my understanding, it doesn't just take any `torch.device` it takes a `str` for device and that `str` *must* be `cuda` or `cpu` (or `xpu` whatever that is), so `mps` won't work. My guess is this is an oversight and will be fixed in PyTorch once they officially support M1.

```
    def __init__(self, device_type : str,
                 dtype : Optional[_dtype] = None,
                 enabled : bool = True,
                 cache_enabled : Optional[bool] = None):
        if torch._jit_internal.is_scripting():
            self._enabled = enabled
            self.device = device_type
            self.fast_dtype = dtype
            # TODO: support get_autocast_gpu/cpu_dtype
            assert dtype is not None
            return
        self.device = device_type
        if self.device == 'cuda':
            self.fast_dtype = torch.get_autocast_gpu_dtype()
        elif self.device == 'cpu':
            self.fast_dtype = torch.get_autocast_cpu_dtype()
        elif self.device == 'xpu':
            self.fast_dtype = torch.xpu.get_autocast_xpu_dtype()  # type: ignore[attr-defined]
        else:
            raise RuntimeError('User specified autocast device_type must be \'cuda\' or \'cpu\'')
```